File: static/cpp/java/src/main/java/com/epi/LongestNondecreasingSubsequenceN2.java
package com.epi;

import java.util.Arrays;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class LongestNondecreasingSubsequenceN2 {
  // @include
  public static int longestNondecreasingSubsequenceLength(List<Integer> A) {
    // maxLength[i] holds the length of the longest nondecreasing subsequence of
    // A[0 : i].
    Integer[] maxLength = new Integer[A.size()];
    Arrays.fill(maxLength, 1);
    for (int i = 1; i < A.size(); ++i) {
      for (int j = 0; j < i; ++j) {
        if (A.get(i) >= A.get(j)) {
          maxLength[i] = Math.max(maxLength[i], maxLength[j] + 1);
        }
      }
    }
    return Collections.max(Arrays.asList(maxLength));
  }
  // @exclude
}

File: static/python/longest_nondecreasing_subsequence_n2.py
# @include
def longest_nondecreasing_subsequence_length(A):
    # max_length[i] holds the length of the longest nondecreasing subsequence
    # of A[:i + 1].
    max_length = [1] * len(A)
    for i in range(1, len(A)):
        max_length[i] = max(1 + max(max_length[j] for j in range(i)
                                    if A[i] >= A[j]), max_length[i])
    return max(max_length)
# @exclude
